### ISSUE TITLE
Laptop Webcam

### ISSUE THREAD

**Issue by jomdolalas** [ultralytics/yolov5#844]
## ❔Question
Hi! Would like to know how you were able to use your webcam as the input? The code
!python detect.py --source 0
does not work for me. I am using a laptop with a webcam installed. Do I need to use an external one?

## Additional context
I also would want to only detect humans/persons

**Comment by glenn-jocher** [comment]
@jomdolalas inference for all sources are shown in https://github.com/ultralytics/yolov5#inference

If you are getting an error, you have environment issues. I'll paste your our default response. 

it appears you may have environment problems. Please ensure you meet all dependency requirements if you are attempting to run YOLOv5 locally.  If in doubt, create a new virtual Python 3.8 environment, clone the latest repo (**code changes daily**), and `pip install -r requirements.txt` again. We also highly recommend using one of our verified environments below. 


## Requirements

**Python 3.8** or later with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies installed, including **`torch>=1.6`**. To install run:
```bash
$ pip install -r requirements.txt
```


## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab Notebook** with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- **Kaggle Notebook** with free GPU: [https://www.kaggle.com/models/ultralytics/yolov5](https://www.kaggle.com/models/ultralytics/yolov5)
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) 
- **Docker Image** https://hub.docker.com/r/ultralytics/yolov5. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) ![Docker Pulls](https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker)


## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are passing. These tests evaluate proper operation of basic YOLOv5 functionality, including training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([test.py](https://github.com/ultralytics/yolov5/blob/master/test.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/models/export.py)) on MacOS, Windows, and Ubuntu.

**Comment by jomdolalas** [comment]
@glenn-jocher Works perfectly now, I just needed to download pytorch 1.6.0. 
After running python detect.py --source 0 locally, I cant seem to close the output window and return to my terminal. Would you have a solution for this?

**Comment by arafmustavi** [comment]
Hi!

I could easily invoke my webcam but I wanted to know if there was any good way to stop the program other than ` ctrl + C `

is there any way that my webcam's video along with detected labels can be stored in any directory?

Thank you!

**Comment by glenn-jocher** [comment]
@arafMustavi the answers to all your questions are in detect.py argparser:
https://github.com/ultralytics/yolov5/blob/e8fc97aa3891f05812d7dfff90ca66d3481bda2c/detect.py#L148-L164

**Comment by Peterson-Repository** [comment]
@arafMustavi

How did you invoke the Webcam for Yolo v5?

**Comment by glenn-jocher** [comment]
@Peterson-Repository you can run webcam inference on a local machine with YOLOv5 using:
```
python detect.py --source 0
```

**Comment by arafmustavi** [comment]
Adding further to @glenn-jocher , if you are using external webcam, try using 1 , 2 , 3 or even -1 after source for running the inferance based on video stream from external source.

**Comment by Peterson-Repository** [comment]
@arafMustavi

is there a way to use custom weights? Looking to use my own .pt file when invoking the camera.

**Comment by glenn-jocher** [comment]
@Peterson-Repository detect.py accepts a variety of arguments including --weights. See argparser for full details.
https://github.com/ultralytics/yolov5/blob/b18ca31ff5412d30ee56eb9c4adbe731f9d2148e/detect.py#L151-L174

**Comment by arafmustavi** [comment]
@Peterson-Repository , Yes there is!

Lets say you have trained a model and the newly trained .pt is under the **custom-trained-model** folder.
use the folowing command:

`python3 detect.py  --weight custom-trained-model/my-trained-model.pt `

The answer @glenn-jocher provided is more generic and it helps you to tweak around all the facilities yolov5 provides throgh their command line. I would suggest you go through the syntax and usage of command line arguments in python and how they are used. It will help a lot to grasp the yolov5's potential.

**Comment by Peterson-Repository** [comment]
Soooo something like

python3 detect.py --source 0 --weights my_custom_weights.pt --conf 0.4


I'm assuming the weights are pulled from the "weights folder", but the "weights folder" only has "download_weights.sh" located in it, not "yolov5s.pt"


i.e. what directory do I drop my custom .pt file?

**Comment by Peterson-Repository** [comment]
Lol, you answered the question before I was done typing.

**Comment by ghost** [comment]
> Hi!
> 
> I could easily invoke my webcam but I wanted to know if there was any good way to stop the program other than `ctrl + C`
> 
> is there any way that my webcam's video along with detected labels can be stored in any directory?
> 
> Thank you!

@arafMustavi  Have you found a way to do this? I put a keyboard exception in the main function but I don't know if this was a good way.

**Comment by glenn-jocher** [comment]
@iandanielsooknanan you can exit any python loop with a `break` statement.

**Comment by RoboflowTeamSat** [comment]
@glenn-jocher I get the error ": cannot connect to X server" when I try connecting to my laptop webcam or an external camera. Any suggestions?

**Comment by glenn-jocher** [comment]
@RoboflowTeamSat this line is supposed to determine if showing the results is viable in your environment:
https://github.com/ultralytics/yolov5/blob/dabad5793a638cba1e5a2bbb878c9b87fe1a14a0/detect.py#L82

You might try forcing this to False. If the error is caused by showing results then this would fix it. If the error is remains then that's between cv2 and your hardware and you should raise your issue on the cv2 repo.

**Comment by akeidn** [comment]
> Hi!
> 
> I could easily invoke my webcam but I wanted to know if there was any good way to stop the program other than `ctrl + C`
> 
> is there any way that my webcam's video along with detected labels can be stored in any directory?
> 
> Thank you!

**Comment by glenn-jocher** [comment]
@akeidn webcam inference is automatically saved to runs/detect/exp#. You can save detections also with:
```
python detect.py --source 0 --save-txt
```

See detect.py argparser for all options, or type `python detect.py --help`

**Comment by youssef12347** [comment]
>  Hi!
>  I could easily invoke my webcam but I wanted to know if there was any good way to stop the program other than `ctrl + C`
>  is there any way that my webcam's video along with detected labels can be stored in any directory?
>  Thank you!

Is there any way to stop the run function by closing the webcam window?

**Comment by nsi1409** [comment]
> > Hi!
> > I could easily invoke my webcam but I wanted to know if there was any good way to stop the program other than `ctrl + C`
> > is there any way that my webcam's video along with detected labels can be stored in any directory?
> > Thank you!
> 
> Is there any way to stop the run function by closing the webcam window?

@youssef12347 

Press the "Q" key on your keyboard while focused on the live video window.

**Comment by aryabhosale** [comment]
> @akeidn webcam inference is automatically saved to runs/detect/exp#. You can save detections also with:
> 
> ```
> python detect.py --source 0 --save-txt
> ```
> 
> See detect.py argparser for all options, or type `python detect.py --help`

i want to use webcam for yolov7 can i use --source 0 for that or i ll get an error. pls send me code for real time testing using webcam (yolov7)
thanks
